### ISSUE TITLE
HashCode for Coordinate in 3D

### ISSUE THREAD

**Issue by JulienDiTria** [locationtech/jts#366]
Is there a reason why the z coordinate of Coordinate class is not used for the hashCode?
even if z = NaN, this works 
public int hashCode() {
    //Algorithm from Effective Java by Joshua Bloch [Jon Aquino]
    int result = 17;
    result = 37 * result + hashCode(x);
    result = 37 * result + hashCode(y);
    result = 37 * result + hashCode(getZ());
    return result;
  }

**Comment by dr-jts** [comment]
I think the rationale was that JTS is essentially a 2D library - the Z coordinate just "comes along for the ride".  So it was thought that hashCode usage would only need to consider the XY ordinates.

Is there a strong reason to change now?

**Comment by JulienDiTria** [comment]
I understand, the only reason I asked is that I m working on two different altitudes for a project and I had to merge 2 hashset of coordinates from different altitude into one. But as only the x and y where taken into account, some Coordinate where not added as they had the same hash. I did implement the hash in 3D but I had to change from a hashset to a hashmap.
For me, it would make sense to compare the Z, as it is available, and therefore adding it to the hash.

**Comment by janruo** [comment]
I agree that it would make sense to include the Z (and M) values for `hashCode()` and `equals()`. Even though JTS is primarily a 2D library, it still contains some 3D algorithms like:
- `org.locationtech.jts.geom.Triangle.area3D()`
- `org.locationtech.jts.geom.Triangle.interpolateZ()`
- `org.locationtech.jts.algorithm.CGAlgorithms3D`
- `org.locationtech.jts.operation.distance3d.Distance3DOp`
- etc...

So clearly the Z value matters and shouldn't be ignored. This is also important because other libraries expect that `hashCode()` and `equals()` are implemented correctly. If two objects are equal and they have the same hash code, they should be interchangeable, but they aren't.

I was hit by this when working on a program that processes triangulated irregular networks. The program reads a TIN from a file and creates a set of Triangle objects. It then uses a third-party universal serializing library to save the set of Triangles efficiently into a cache file for later use, because reading the original file format is slow and expensive. The program then uses the set of Triangles to do some processing, mostly 2D stuff by flattening the TIN into a Polygon. But one of the features is a 3D cross section, which needs the Z value of the Coordinates. It worked fine, but only when the TIN was freshly loaded from the original file. When the TIN was deserialized from the cache, it subtly gave incorrect results. After a while of debugging, it turned out that the serializer did some optimization by calling `equals()` and `hashCode()` to determine whether two objects are the same (which is pretty reasonable). But because the Coordinate's implementation doesn't care about the Z value, the deserializer restored the TIN with incorrect height data if it encountered more than one coordinate with the same X and Y values but different Z values. I resolved this issue by using an older version of the serializer that checked for reference equality, not `equals()`.

**Comment by dr-jts** [comment]
I mostly agree with this, but one that gives me pause is that most JTS algorithms are 2D only, and there may be some that rely on the 2D-only hashcode to work correctly.  I can't offhand think of any that are affected, but there probably are some.  

Not a reason not to change this, but it does make it potentially a more complicated fix.  (And introduces a further complexity going forward).
